reactor-core/src/main/java/reactor/core/publisher/SinksSpecs.java

```diff
@@ -78,16 +77,25 @@ public void emitComplete() {
 
 	@Override
 	public final Emission tryEmitComplete() {
-		if (done) {
-			return Sinks.Emission.FAIL_TERMINATED;
+		Thread currentTread = Thread.currentThread();
+		if (WIP.get(this) == 0 && WIP.compareAndSet(this, 0, 1)) {
+			LOCKED_AT.lazySet(this, currentTread);
```

Comment:
note to reviewers:

lazy write should(?) be fine here because only the current thread really cares about the value

Reply:
as discussed offband, would actually benefit from a detailed comment because `lazySet` has subtle implications.

the winning thread `A` update could be invisible to thread `B` performing `LOCKED_AT.get` below, but that is ok as long as the conditional is `!= currentThread`. What can happen at the `get`:
 - the A update is visible ✅ 
 - the A update is not visible, `null` is seen
 - the A update is not visible, `C` is seen (any other old winning thread)

But having an outdated `B` value is not possible because to reach that point, since we're in thread B, it must have executed the `compareAndSet` which would have loaded the update from `A`.